# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/layers/NativeLayerCA.h
# Commit: 362b1e430789
# Full Hash: 362b1e430789506fb706b6eea448326d84f80c24
# Author: Brad Werth <bwerth@mozilla.com>
# Date: 2022-07-21 03:47:38
# Regressor Bug: 1737682
# File Overlap Count: 1
# Description:
#   Bug 1737682 Part 1: Log whether or not we are hitting the detached state. r=mstange
#   
#   There are 10 enum cases that we track internally, one of which isn't
#   emitted as telemetry. These cases are 0-indexed to match the enum values.
#   
# ==============================================================================

diff -r ed9dddbf5496 -r 362b1e430789 gfx/layers/NativeLayerCA.h
--- a/gfx/layers/NativeLayerCA.h	Wed Jul 20 18:40:56 2022 +0000
+++ b/gfx/layers/NativeLayerCA.h	Wed Jul 20 18:48:28 2022 +0000
@@ -42,6 +42,21 @@
 class NativeLayerRootSnapshotterCA;
 class SurfacePoolHandleCA;
 
+enum class VideoLowPowerType {
+  // These must be kept synchronized with the telemetry histogram enums.
+  NotVideo,           // Never emitted as telemetry. No video is visible.
+  Success,            // As best we can tell, we are in the "detached",
+                      // low-power compositing mode.
+  FailMultipleVideo,  // There is more than one video visible.
+  FailWindowed,       // The window is not fullscreen.
+  FailOverlaid,       // Something is on top of the video (likely captions).
+  FailBacking,        // The layer behind the video is not full-coverage black.
+  FailMacOSVersion,   // macOS version does not meet requirements.
+  FailPref,           // Pref is not set.
+  FailSurface,        // Surface is not eligible.
+  FailEnqueue,        // Enqueueing the video didn't work.
+};
+
 // NativeLayerRootCA is the CoreAnimation implementation of the NativeLayerRoot
 // interface. A NativeLayerRootCA is created by the widget around an existing
 // CALayer with a call to CreateForCALayer - this CALayer is the root of the
@@ -122,7 +137,8 @@
       gfx::DeviceColor aColor) override;
 
   void SetWindowIsFullscreen(bool aFullscreen);
-  void NoteMouseMoveAtTime(const TimeStamp& aTime);
+
+  VideoLowPowerType CheckVideoLowPower();
 
  protected:
   explicit NativeLayerRootCA(CALayer* aLayer);
@@ -164,6 +180,12 @@
   // Updated by the layer's view's window to match the fullscreen state
   // of that window.
   bool mWindowIsFullscreen = false;
+
+  // How many times have we committed since the last time we emitted
+  // telemetry?
+  unsigned int mTelemetryCommitCount = 0;
+  static const unsigned int TELEMETRY_COMMIT_PERIOD =
+      600;  // 10 seconds at 60fps
 };
 
 class RenderSourceNLRS;